subsys/dfu/dfu_multi_image/src/dfu_multi_image.c

```diff
@@ -176,6 +255,169 @@ static int process_current_item(const uint8_t *chunk, size_t chunk_size)
+			return rc;
+		}
+
+		ctx.cur_offset = ctx.cur_item_size;
+
+		select_next_image();
+	}
+
+	if (ctx.buffer != NULL && strcmp(key, IMAGE_FINISHED_SETTING_NAME) == 0) {
+		/* Restore the cbor header from settings if available */
+		uint8_t finished_image_no = 0;
+
+		if (len_rd != sizeof(finished_image_no)) {
+			/* Should never get here */
+			return -EINVAL;
+		}
+
+		len = read_cb(cb_arg, &finished_image_no, sizeof(finished_image_no));
+
+		if (len < 0) {
+			LOG_ERR("Can't read finished image number from storage");
+			return len;
+		}
+
+		if (ctx.max_loaded_finished_image_no < finished_image_no) {
+			ctx.max_loaded_finished_image_no = finished_image_no;
+		}
+	}
+
+	return 0;
+}
+
+SETTINGS_STATIC_HANDLER_DEFINE(dfu_multi_image, MODULE_NAME, NULL, settings_set,
+			       NULL, NULL);
+
+static int load_saved_progress(void)
+{
+	int err = 0;
+
+	LOG_INF("Loading saved progress from NVM");
+
+	ctx.max_loaded_finished_image_no = -1;
+
+	/* settings_subsys_init is idempotent so this is safe to do. */
+	err = settings_subsys_init();
+	if (err) {
+		LOG_ERR("settings_subsys_init failed (err %d)", err);
+		return err;
+	}
+
+	/* Load settings from NVM. If the update was interrupted, the context will be restored */
+	err = settings_load_subtree(MODULE_NAME);
+	if (err) {
+		LOG_ERR("settings_load failed (err %d)", err);
+		return err;
+	}
+
+	if (ctx.cur_image_no == IMAGE_NO_FIXED_HEADER) {
+		LOG_INF("No saved progress found");
+		/* Progress was not stored in settings, start from the beginning */
+		ctx.saved_progress_loaded = true;
+		return 0;
+	}
+
+	if (ctx.cur_image_no < 0) {
+		/* Should never get here */
+		return -EFAULT;
+	}
+
+	while (ctx.cur_image_no < ctx.header.image_count) {
+		const struct dfu_image_writer *writer = current_image_writer();
+
+		if (writer == NULL) {
+			LOG_WRN("No writer registered for image %d!", ctx.cur_image_no);
+			err = -EPIPE;
+			break;
+		}
+
+		if (writer->offset == NULL) {
+			LOG_WRN("No function to read offset registered for image %d!",
+				ctx.cur_image_no);
+			err = -EPIPE;
+			break;
+		}
+
+		err = writer->open(writer->image_id,
+				   ctx.header.images[ctx.cur_image_no].size);
+
+		if (!err) {
+			ctx.cur_item_opened = true;
+			err = writer->offset(&ctx.cur_item_offset);
+		}
+
+		if (!err) {
+			if (ctx.cur_image_no <= ctx.max_loaded_finished_image_no) {
+				/* This image was fully written before power loss/reset */
+				ctx.cur_offset += ctx.cur_item_size;
+			} else {
+				ctx.cur_offset += ctx.cur_item_offset;
+
+				if (ctx.cur_item_offset < ctx.cur_item_size) {
+					/*
+					 * Writing to the current image is not finished.
+					 * Resume writing from the current offset.
+					 */
+					break;
+				}
+			}
+			/*
+			 * The current image is already fully written.
+			 * Close the writer and move to the next image.
+			 */
+			err = writer->close(true);
+			ctx.cur_item_opened = false;
+		} else {
+			LOG_ERR("Failed to restore image %d offset", ctx.cur_image_no);
```

Comment:
Perhaps we need a Kconfig determining what to do if loading the progress has failed. For many application protocols it should be fine to start from the beginning of an image (and re-request the same data) or from scratch.